core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -492,19 +542,61 @@ class LogTest extends JUnitSuite {
     val config = LogConfig(logProps)
     var log = new Log(logDir, config, recoveryPoint = 0L, time.scheduler, time)
     for(i <- 0 until numMessages)
-      log.append(TestUtils.singleMessageSet(TestUtils.randomBytes(10)))
+      log.append(TestUtils.singleMessageSet(payload = TestUtils.randomBytes(10), timestamp = time.milliseconds + i * 10))
     val indexFiles = log.logSegments.map(_.index.file)
+    val timeIndexFiles = log.logSegments.map(_.timeIndex.file)
     log.close()
 
     // delete all the index files
     indexFiles.foreach(_.delete())
+    timeIndexFiles.foreach(_.delete())
 
     // reopen the log
     log = new Log(logDir, config, recoveryPoint = 0L, time.scheduler, time)
     assertEquals("Should have %d messages when log is reopened".format(numMessages), numMessages, log.logEndOffset)
-    for(i <- 0 until numMessages)
+    assertTrue("The index should have been rebuilt", log.logSegments.head.index.entries > 0)
+    assertTrue("The time index should have been rebuilt", log.logSegments.head.timeIndex.entries > 0)
+    for(i <- 0 until numMessages) {
       assertEquals(i, log.read(i, 100, None).messageSet.head.offset)
+      if (i == 0)
+        assertEquals(log.logSegments.head.baseOffset, log.fetchOffsetsByTimestamp(time.milliseconds + i * 10))
+      else
+        assertEquals(i, log.fetchOffsetsByTimestamp(time.milliseconds + i * 10))
+    }
+    log.close()
+  }
+
+  /**
+   * Test that if message format version is set to before 0.10.0, the time index should only contain the one entry
+   * point to the base offset of next segment.
+   */
+  @Test
+  def testRebuildTimeIndexForOldMessages() {
+    val numMessages = 200
+    val segmentSize = 200
+    val msgPerSeg = segmentSize / TestUtils.singleMessageSet(TestUtils.randomBytes(10)).sizeInBytes
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, segmentSize: java.lang.Integer)
+    logProps.put(LogConfig.IndexIntervalBytesProp, 1: java.lang.Integer)
+
+    val config = LogConfig(logProps)
+    var log = new Log(logDir, config, recoveryPoint = 0L, time.scheduler, time)
+    for(i <- 0 until numMessages)
+      log.append(TestUtils.singleMessageSet(payload = TestUtils.randomBytes(10), timestamp = time.milliseconds + i * 10))
```

Comment:
It does not really matter whether we append message in 0.9.0 or 0.10.0 format because the log segment will do down conversion when the message is appended if the format is 0.10.0